[WIP] Anomaly likelihood

breznak opened the pull request:
Early work, no need to review yet. 

TODO:
- [ ] integrate into nupic's cmake build system
- [ ] port tests from tests/unit/.... in nupic-PY
- [ ] create a header file (at the end, when everything is cleaned up) 
- [ ] integrate into `Anomaly` class

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,158 @@
+  Real32 stdev;
+};
+
+struct DataPoint {
```
I'd consider eliminating this struct and using passing all 3 required arguments to the respective methods. To keep the API same as in the python version.

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,158 @@
+    bool timestampSeted;
+};
+
+class MovingAverage {
```
no need for this. Remove and please use the impl from `nupic/utils/MovingAverage`

breznak commented on src/nupic/algorithms/reference_anomaly_likelihood.py:
```diff
@@ -0,0 +1,595 @@
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+# TODO remove this file after porting!
```
remove this file in the end

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,158 @@
+        probationaryPeriod = learningPeriod+estimationSamples;
+        iteration = 0;
+    }
+    AnomalyLikelihood(int learningPeriod, int estimationSamples, int historicWindowSize, int reestimationPeriod) :
```
use `UInt` (nupic's unsigned int) , otherwise ok

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+    }
+
+    
+    Real anomalyProbability(Real rawValue, Real anomalyScore, int timestamp=-1) { //TODO "rawValue" could be typed to T rawValue; //TODO is rawValue ever used?
```
TODO: `rawValue` could be templated (T), not a priority now. 
FIXME: is the rawValue ever used? Does not seem so in my implementation.

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+      if ( this->iteration == 0 || (this->iteration % this->reestimationPeriod) == 0  || this->distribution.name == "unknown" ) {
+
+        auto numSkipRecords = this->calcSkipRecords_(this->iteration, this->runningAverageAnomalies.capacity(), this->learningPeriod); //FIXME this erase (numSkipRecords) is a problem when we use sliding window (as opposed to vector)! - should we skip only once on beginning, or on each call of this fn?
+        estimateAnomalyLikelihoods(anomalies, numSkipRecords);  // called to update this->distribution;  
```
FIXME this erase (numSkipRecords) - should we skip only once on beginning, or on each call of this fn?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+    boost::circular_buffer<Real> runningLikelihoods; // sliding window of the likelihoods
+    boost::circular_buffer<Real> runningRawAnomalyScores; 
+    boost::circular_buffer<Real> runningAverageAnomalies; //sliding window of running averages of anomaly scores
+    boost::circular_buffer<Real> runningRawValues; // sliding window of the raw values
```
TODO: actually, it seems only a few (runningAverageAnomalies) are used, should all other be deleted, or kept for "clarity"?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+}
+
+
+DistributionParams estimateNormal(vector<Real> sampleData, bool performLowerBoundCheck=true) {
```
TODO: should I create a class DistributionParams (currently a struct) and check/compute all this info in the class, instead of this method?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+
+Real compute_mean(vector<Real> v) const { //TODO do we have a (more comp. stable) implementation of mean/variance?
+    Real sum = std::accumulate(v.begin(), v.end(), 0.0);
+    return sum / v.size();
```
TODO is there an implementation of mean/variance to reuse? instead of writing it here...

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+
+  for (Real v: likelihoods) { 
+   
+    if (v <= redThreshold) { //TODO review the IFs
```
TODO please review after me

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+}
+
+
+vector<Real> estimateAnomalyLikelihoods(vector<Real> anomalyScores, UInt skipRecords=0, UInt verbosity=0) { //FIXME averagingWindow not used, I guess it's not a sliding window, but aggregating window (discrete steps)!
```
FIXME `averagingWindow=10` not used, I guess it's not a sliding window, but aggregating window (discrete steps)! 
...Now I use it in `averagedAnomalies (MovingAverage)` to compute a small sliding windows/averages - it mimics `_anomalyScoreMovingAverage()` from py.

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+  }
+
+  assert(anomalyScores.size() >= 0); // "Must have at least one anomalyScore"
+  auto dataValues = anomalyScores; //FIXME the "data" should be anomaly scores, or raw values? 
```
what should the `dataValues` be? raw values, raw anomaly scores, aggreggated anom scores (current), likelihoods,...?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+  
+    /* HACK ALERT! The CLA model currently does not handle constant metric values //TODO can this be removed now? 4/2017
+     very well (time of day encoder changes sometimes lead to unstable SDR's
+     even though the metric is constant). Until this is resolved, we explicitly
```
is the HACK still needed?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+
+vector<Real> circularBufferToVector(boost::circular_buffer<Real> cb) const {
+  cb.linearize();
+  auto d1 = cb.array_one();
```
I need to get data as a vector from the sliding window, for the `boost::circular_buffer` I need this hack. This is what I was mentioning in #1276

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,592 @@
+  // Filter the likelihood values. First we prepend the historical likelihoods
+  // to the current set. Then we filter the values.  We peel off the likelihoods
+  // to return and the last windowSize values to store for later.
+  UInt toCrop = min((unsigned long)this->averagedAnomaly.getMaxSize(), runningLikelihoods.size());
```
TODO: review, if this should be current  `this->averagedAnomaly.getMaxSize()`, 
or `circular_buffer.capacity()` from this->runningAveragedAnomalies?

breznak replied:
still valid

breznak commented on src/nupic/utils/MovingAverage.hpp:
```diff
@@ -43,6 +43,7 @@ namespace nupic
       Real32 getCurrentAvg() const;
       Real32 compute(Real32 newValue);
       Real32 getTotal() const;
+      UInt getMaxSize() const;
```
rename to `capacity` (?)

rhyolight commented:
@breznak Can you investigate the Travis and Appveyor failures?

breznak commented:
> Can you investigate the Travis and Appveyor failures? 

Sure, but first I'd like a high level review (the code compiles for me on linux, i checked and those will be platform/version nuances like "Werror on comparison of unsigned int and unsigned long", etc

breznak commented:
*) commit e01c7fd causes a linking problem (in the swig `.i` files?) Is there a need to edit the wrappers? 

```
[ 79%] Built target _engine_internal
/tmp/cclyBd1I.ltrans30.ltrans.o: In function `nupic::algorithms::anomaly::Anomaly::compute(std::vector<unsigned int, std::allocator<unsigned int> > const&, std::vector<unsigned int, std::allocator<unsigned int> > const&, double, unsigned int) [clone .constprop.1138]':
<artificial>:(.text+0x4ae): undefined reference to `nupic::algorithms::anomaly::AnomalyLikelihood::anomalyProbability(float, float, int)'
/tmp/cclyBd1I.ltrans30.ltrans.o: In function `nupic::algorithms::anomaly::Anomaly::Anomaly(unsigned int, nupic::algorithms::anomaly::AnomalyMode, float) [clone .constprop.1137]':
<artificial>:(.text+0x4fa): undefined reference to `nupic::algorithms::anomaly::AnomalyLikelihood::AnomalyLikelihood(unsigned int, unsigned int, unsigned int, unsigned int, unsigned int)'
```

breznak commented:
> Can you investigate the Travis and Appveyor failures? 

@rhyolight fixed. 

Now I'm looking for someone to review the PR (high level). 
And help me fix [what broke in integration with Anomaly](https://github.com/numenta/nupic.core/pull/1268#issuecomment-294787667)

breznak commented:
@rhyolight where are you, guys? :duck:

breznak commented:
@Thanh-Binh fixed in the last commit. Thank you for reporting!

rhyolight commented:
Looks like there is still one failing test on Windows.

rhyolight commented:
@breznak Sorry I took last week off. I would like @scottpurdy or @mrcslws to review this, but I don't know if they have time right now. This is a big update and deserves some attention before being merged.

Thanh-Binh commented:
I test this code with some applications, it seems to be provided understandable results

rhyolight commented:
@breznak I think that @mrcslws can review this in our next sprint, which starts Friday and lasts 2 weeks. We have scheduled the work.

breznak commented:
@rhyolight No, problem, hope you enjoyed your holiday :) 
Perfect! Thanks for putting this on schedule. 

In the meanwhile I'll try to fix the portability issues in CI.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.hpp:
```diff
@@ -0,0 +1,171 @@
+
+static Real compute_var(vector<Real> v, Real mean)  {
+    Real sq_sum = std::inner_product(v.begin(), v.end(), v.begin(), 0.0);
+    return sq_sum / v.size() - mean * mean;
```
I think it'd be better to use explicit parentheses to show order of operations. This code works, but it will be more readable and less scary with parens.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,505 @@
+     detect and handle completely flat metric values by reporting them as not
+     anomalous.
+    */
+      auto metricDistribution = estimateNormal(dataValues, false);
```
You're passing in the anomaly scores. You should pass in the raw values. This is detecting the case where the raw values are repeatedly the same.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,505 @@
+    Real likelihood = 0.5f;
+
+    if (timestamp == -1) {
+      timestamp = this->iteration;
```
All of these `this->member`s should be `member_`. Private members should have a trailing underscore, and we don't need to do `this->`.

breznak replied:
Why is there a preference to a (local) naming convention (localName_) to a self explanatory `this->someVar` ?

mrcslws replied:
Trailing underscores show that it's private. And it's less verbose.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,505 @@
+    }
+
+    
+Real AnomalyLikelihood::anomalyProbability(Real rawValue, Real anomalyScore, int timestamp) { //TODO "rawValue" could be typed to T rawValue; //TODO is rawValue ever used?
```
You're correct that it's weird that `rawValue` is a specific type. Really it's weird that this class even receives the raw values. It's because of the big flatline hack. I think ultimately that hack should be removed from the anomaly likelihood code and the caller should handle it. But that would involve changing nupic. Maybe the best thing is to remove the hack in the C++ code and just document that this code doesn't offer that hack "feature".

mrcslws commented on src/nupic/algorithms/Anomaly.hpp:
```diff
@@ -112,10 +113,16 @@ namespace nupic
         AnomalyMode mode_;
         Real32 binaryThreshold_;
         std::unique_ptr<nupic::util::MovingAverage> movingAverage_;
+        AnomalyLikelihood likelihood_; //TODO which params/how pass them to constructor?
```
This should be a pointer. Then we should create it in the constructor, but only for AnomalyModes that use it. Then delete it in the destructor.

breznak replied:
Is it worth adding the code to instantiate only if needed and destruct, rather than just make an instance and use it when needed?

breznak replied:
> This approach, having members that are only used in some configurations, is not generally seen as good C++ design. 

So let's make it right! 
How about: 
- `Anomaly` interface with `virtual Real compute(...)`
- `AnomalyRaw : Anomaly` which is basically this class in the current state (w/o the likelihood_ field)
- `AnomalyLikelihood: AnomalyRaw` which will override the compute() ? 
- we could then get rid of the `AnomalyMode` is it would be inferred by the class inheritance
- I just don't know how to implement the mode `WRIGHTED` which is `raw()*likelihood()`, but maybe that can be left to the client side, and we'll only provide raw & likelihood here. 
@mrcslws shall I proceed with that?

breznak replied:
The problem would be with different parameters for Raw/Likelihood. Or we'd have to use redundant params: `virtual Real compute(vector active, vector predicted, Real rawAnomalyScore, int timestamp)`. 
Also, I'd like to rename `anomalyProbability()` to `compute()` ,which will diverge from the python impl (we can leave comment), but align with the c++ codebase.

mrcslws commented on src/nupic/algorithms/Anomaly.hpp:
```diff
@@ -112,10 +113,16 @@ namespace nupic
                        Real64 inputValue=0, UInt timestamp=0);
 
+//helper
+        void setLikelihood(AnomalyLikelihood newInstance) {
```
I don't think we need this method.

breznak replied:
Currently not all parameters in Likelihood can be set in Anomaly, so this helps by allowing to fine-tune and instantiate Likelihood as needed and then just plug it in here. (?)

breznak commented on src/nupic/algorithms/AnomalyLikelihood.hpp:
```diff
@@ -0,0 +1,171 @@
+    DistributionParams nullDistribution() const;
+
+    //private static methods
+static Real  computeLogLikelihood(Real likelihood)  {
```
Just a quick note, why should this be public? The python file has only 2 high-level fuctions: constructor and anomalyProbabilities() (=compute()); the rest is described as "helper methods", so I thought I'd hide that implementation into `private:` (?)

mrcslws replied:
It's intended for public use, e.g. for visualization. In the Python, private methods are denoted by a leading underscore, e.g. `_calcSkipRecords`.

Also note that it doesn't have any callers in `anomaly_likelihood.py`.

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,505 @@
+    averagedAnomaly(aggregationWindow) {
+        iteration = 0;
+        probationaryPeriod = learningPeriod+estimationSamples;
+        assert(historicWindowSize >= estimationSamples); // cerr << "estimationSamples exceeds historicWindowSize";
```
@rhyolight what is the difference between NTA_CHECK/ASSERT? The check stays, while assert can be optimized away?

mrcslws replied:
`NTA_ASSERT`s are removed in release builds. `NTA_CHECK`s are always included.

breznak commented:
@mrcslws can you please confirm that bae48b2 is how you intended to remove the "hack"?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,505 @@
+      return 0.5f;
+    } //else {
+
+    auto anomalies = circularBufferToVector(this->runningAverageAnomalies); 
```
I'll avoid this when we use code from #1276

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,505 @@
+  this->runningLikelihoods.insert(runningLikelihoods.end() - toCrop, likelihoods.begin(),likelihoods.end()); //append & crop
+  assert(this->runningLikelihoods.size() <= this->averagedAnomaly.getMaxSize());
+
+  auto filteredLikelihoods = filterLikelihoods_(likelihoods);
```
Isn't this another use-case for `auto`? It allows us to change the return-type in the method easity (as long as it's compatible)...

mrcslws replied:
It's better to optimize for readability. Changing return-types is rare and changing the code is easy.

breznak commented:
Thank you very much for the first review, @mrcslws !! 

I did most of your suggestions, commented on the others. 

From your review: 

> ... I wasn't totally clear what parts were ready for review.

All parts should be working and are ready for your review. 

> It'd be best if you find a way to stop making a full copy of the runningAverageAnomalies on every iteration.

Yes, we'll be able to get rid of `boost::circular_buffer` and replace it with our own implementation, which will provide the vector right away; That will be done in #1276 and used here. 

> The "HACK ALERT" part is complicated. I don't think we should translate that hack into C++, especially since it's so much harder to implement in C++ (it would require templates). We should get rid of the hack...

Implemented and removed in bae48b2. 

From my review: 

> only raw anomaly score could be used as input (no raw value, timestamp needed)?

Ok, `rawValue` removed (commit above); it seems we could also remove the timestamp; or make the code use it, as `iteration` is actually used instead. 

> the anomalies/likelihoods/averaged-/aggregated-anomalies in PY reference are quite confusing, please help me double check I'm passing correct data to the methods.

This still holds. 

- In your next review, please focus on TODOs, FIXMEs in the code, thank you!

breznak commented:
@mrcslws can we continue the review process, please?

mrcslws commented:
I've been focusing on https://github.com/numenta/nupic.core/pull/1277, which this change depends on.

I'm heads-down preparing a presentation. I'll do another round of reviews on **both** PRs this week, but I might not get to it until Friday.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.hpp:
```diff
@@ -0,0 +1,261 @@
+}
+
+    //private static methods
+static Real  computeLogLikelihood(Real likelihood)  {
```
As mentioned before, this should be a public static method.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.hpp:
```diff
@@ -0,0 +1,261 @@
+    return log(1.0000000001f - likelihood) / -23.02585084720009f;
+}
+
+static UInt calcSkipRecords_(UInt numIngested, UInt windowSize, UInt learningPeriod)  {
```
This can be a function in the cpp file. It doesn't need to be a method in the header.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.hpp:
```diff
@@ -0,0 +1,261 @@
+
+    DistributionParams distribution ={ "unknown", 0.0, 0.0, 0.0};
+    UInt probationaryPeriod;
+    UInt iteration;
```
Add a trailing underscore to all of these private member variables.

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.hpp:
```diff
@@ -0,0 +1,261 @@
+}
+
+
+static std::vector<Real> circularBufferToVector(boost::circular_buffer<Real> cb) {
```
This shouldn't be in the header. (I know this will disappear when the `MovingWindow` is added, but I'm saying this again for completeness)

mrcslws commented on src/nupic/algorithms/AnomalyLikelihood.hpp:
```diff
@@ -0,0 +1,261 @@
+  samples > x and for values below the mean, the probability of getting
+  samples < x. This is the Q-function: the tail probability of the normal distribution.
+  **/
+    Real32 tailProbability(Real32 x) const;
```
If you intend this to be a private method, it should have a trailing underscore.

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,340 @@
+    }
+
+    
+Real AnomalyLikelihood::anomalyProbability(Real anomalyScore, int timestamp) {  //FIXME even timestamp is not really used, remove too? 
```
What about the timestamp, should I remove it?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,340 @@
+    auto anomalies = circularBufferToVector(this->runningAverageAnomalies_); 
+    
+      // On a rolling basis we re-estimate the distribution
+      if ( this->iteration_ == 0 || (this->iteration_ % reestimationPeriod_) == 0  || distribution_.name == "unknown" ) {
```
...or maybe utilize the timestamp here and replace with it the current iteration_ ?

breznak replied:
Actually, I'm starting to like this approach. It would allow us to treat the parameters (`reestimationPeriod`, etc) as time (in units of timestamp) if `timestamp` is provided; or as iterations if it's not provided (current default).

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,340 @@
+     find that such low anomaly means can happen, but then the slightest blip
+     of anomaly score can cause the likelihood to jump up to red.
+     */
+    if (params.mean < THRESHOLD_MEAN) { //TODO make these magic numbers a constructor parameter, or at least a const variable
```
Should I move it to constructor with default values?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,340 @@
+  :returns: A new list of floats likelihoods containing the filtered values.
+  **/
+static vector<Real> filterLikelihoods_(vector<Real> likelihoods, Real redThreshold=0.99999, Real yellowThreshold=0.999){ //TODO make the redThreshold params of AnomalyLikelihood constructor() 
+  redThreshold    = 1.0 - redThreshold;  //TODO maybe we could use the true meaning already in the parameters
```
What about this?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,340 @@
+  vector<Real> likelihoods;
+  likelihoods.reserve(runningAverageAnomalies_.size()); 
+  for (auto newAverage : runningAverageAnomalies_) { //TODO we could use transform() here (? or would it be less clear?) 
+    likelihoods.push_back(tailProbability_(newAverage)); 
```
Use `transform()` ?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,340 @@
+  UInt toCrop = min((UInt)this->averagedAnomaly_.getMaxWindowSize(), (UInt)runningLikelihoods_.size());
+  this->runningLikelihoods_.insert(runningLikelihoods_.end() - toCrop, likelihoods.begin(),likelihoods.end()); //append & crop
+  cerr << "FUU " << this->runningLikelihoods_.size() << "  <= " << this->averagedAnomaly_.getMaxWindowSize() << endl; 
+//!  NTA_CHECK(this->runningLikelihoods_.size() <= this->averagedAnomaly_.getMaxWindowSize()); //FIXME fix this check, returns ~"14xx <= 10"
```
FIXME: is this check correct? If so, why is it failing?

breznak commented on src/nupic/algorithms/AnomalyLikelihood.cpp:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  NTA_CHECK(anomalyScores.size() > 0); // "Must have at least one anomalyScore"
+  auto dataValues = anomalyScores; //FIXME the "data" should be anomaly scores, or raw values? 
```
This comment can be removed, I assume we deal with anomaly scores only here, right?

mrcslws commented:
Some of our comments criss-crossed. A couple of my comments got sent yesterday while most of them got sent today. Github makes it hard to see everything that has been said. If you can give an itemized list of open questions as a normal PR comment (not as diff comments), that would be really helpful.

breznak commented:
> Some of our comments criss-crossed. A couple of my comments got sent yesterday while most of them got sent today.  

I think I got all the comments from yesterday (~18 hrs ago) and 1 comment from today (~3hrs ago). 

This is a summary of the most important questions for me: 
> - implemented the formatting changes you've suggested in the last review
- I'm waiting for your decision how the better deal with Anomaly in #1268 (comment)
- an interesting issue is with honoring the timestamp, as in #1268 (comment)
- Can you please address the standing TODOs/FIXMEs I've highlighted in the review below?
- consider reset() for Likelihood, #1267 (comment)

About to today's comment:

> ..remove timestamp... 

Actually, I'd like to do the opposite and make the functionality of likelihood account for time, as seen in the next commit

breznak commented:
One more thing. I'd like to discuss __aggregation__ . 
There's a parameter `aggregationWindow`(=5), 
- in Python I think this reduces data for use in `updateAnomalies()` from, say, `25`(/5) => 4. 
- I've mistakenly implemented this as `running average`(of size 5), so I'd get 25 values (smoothed by 5). 
  - computational complexity aside, this could actually be a better approach: the statistics is ~same, you get more detailed data, but still get some rounding (smoothing). 

- Overall, I'd like to propose to __remove__ this functionality from the C++ (maybe from Py too?), as it
  - adds hidden complexity
  - can be easily implemented by the client( `auto a5 = anomalyRaw(with window=5).compute(); if(iter++%5==0) auto score = likelihood.compute(a5);` )
   - is conceptually not optimal (even wrong). If you need to do aggregation, do it on the lower levels (input layer), as it also affects TM (which is used to Likelihoods). 

From the comments I can guess the example for which this was developed is 
> HotGym consumption: 5min data, cycle 1week, we want to predict&detect hourly changes

So there was aggregation used to condense {12x5min} => 1hr. 

TL;DR: aggregation should be done (and is supported) on lower levels (input), not in Likelihood. Remove it from there. 

@mrcslws @subutai

breznak commented:
https://travis-ci.org/numenta/nupic.core/builds/234448366#L5488 

Need to fix this, seems to be passing on my local machine

subutai commented:
> One more thing. I'd like to discuss aggregation .
> There's a parameter aggregationWindow(=5),
> 
> in Python I think this reduces data for use in updateAnomalies() from, say, 25(/5) => 4.
> 
> I've mistakenly implemented this as running average(of size 5), so I'd get 25 values (smoothed by 5).

There's no aggregationWindow parameter in Python, but there is averagingWindow (defaults to 10). It implements a running average with that window size.    I've tested other values - overall it's not super sensitive to it, but 10 works well.  I'm not sure why we would remove the feature - it is currently required.

breznak commented:
>  but there is averagingWindow (defaults to 10). It implements a running average with that window size. 

Perfect, then that's what I have implemented here. And my comment about inappropriate aggregation does not apply. 
Thanks Subutai

breznak commented:
@rhyolight I've closed the fixes, small PRs here and moved them to https://github.com/htm-community/nupic.cpp/pulls 
What about these 2 remaining? It is quite a big feature, do you want to get it into the official version, or should I close it here, and also continue in the community?

rhyolight commented:
We just don't have the people to do the PR review right now. I think you should close and focus on community fork.
